File: bin/report-high-dist-candidates.R
#!/usr/bin/env Rscript
args = commandArgs(trailingOnly=TRUE)

# load necessary packages
require(tidyverse)
require(arrow)
require(Biostrings)

# set just-in-time compilation setting so that top-level loops are JIT compiled
compiler::enableJIT(3)

# Bringing in NCBI metadata
metadata <- read_ipc_file(args[1])

# Defining strictness level
if (args[2] == "strict") {
  stringency <- 995
} else if (args[2] == "intermediate") {
  stringency <- 990
} else if (args[2] == "lenient") {
  stringency <- 980
} else {
  stringency <- 995
}

# create function to process metadata
process_meta <- compiler::cmpfun(function(yearmonths, metadata){
  
  # make an empty data frame to store high distance metadata 
  high_dist_meta <- matrix(ncol = ncol(metadata), nrow = 0)
  colnames(high_dist_meta) <- colnames(metadata)
  high_dist_meta <- as.data.frame(high_dist_meta)
  
  # looping through each year-month and adding the highest distance cluster to 
  # a final table of hits
  first_yearmonth = yearmonths[1]
  for (i in yearmonths){
    
    # read in distance matrix
    distmat = read_csv(paste(i, "-dist-matrix.csv", sep = ""),
                       trim_ws = TRUE, show_col_types = FALSE)
    
    # define high-distance accession
    distmat$Distance_Score <- vapply(2:ncol(distmat), FUN = 
                                       function(j){
                                         
                                         distances <- as.numeric(distmat[,j])
                                         distances <- distances[distances!=0]
                                         new_sum <- sum(distances)
                                         
                                         return(new_sum)
                                         
                                       }, numeric(1))
    distmat <- distmat[order(distmat$Sequence_Name),] ; rownames(distmat) <- NULL
    
    # collating cluster metadata with distances
    cluster_table <- read_tsv(paste(i, "-clusters.uc", sep = ""), 
                              col_names = FALSE, trim_ws = TRUE, show_col_types = FALSE)
    if (nrow(distmat)==1){
      cluster_table <- cluster_table[cluster_table$V9==distmat$Sequence_Name[1],]
    }
    centroids <- cluster_table[cluster_table$V1=="C",]
    centroids <- centroids[order(centroids$V9),] ; rownames(centroids) <- NULL
    stopifnot(nrow(distmat) == nrow(centroids))
    stopifnot(unique(distmat$Sequence_Name == centroids$V9) == TRUE)
    distmat$Cluster <- 0
    distmat$Cluster_Size <- centroids$V3
    distmat <- distmat[order(distmat$Distance_Score, decreasing = T),] ; rownames(distmat) <- NULL
    distmat$Month <- i 
    distmat$Cluster <- centroids$V2
    
    # new cluster table
    all_seqs <- distmat[, 
                        c("Sequence_Name", "Distance_Score", 
                          "Cluster_Size", "Month", "Cluster")]
    
    # adding all cluster sequences
    if ("H" %in% cluster_table$V1){
      hits <- cluster_table[cluster_table$V1 == "H",]
      for (j in 1:nrow(hits)){
        
        # record accession and the centroid for that accession
        accession <- hits$V9[j]
        centroid <- hits$V10[j]
        distance <- all_seqs[all_seqs$Sequence_Name==centroid, "Distance_Score"]
        cluster_size <- all_seqs[all_seqs$Sequence_Name==centroid, "Cluster_Size"]
        month <- i
        cluster <- all_seqs[all_seqs$Sequence_Name==centroid, "Cluster"]
        
        # construct a new row for the all_seqs dataframe
        new_row <- c(accession, distance, cluster_size, month, cluster)
        
        # bind to table
        all_seqs <- rbind(all_seqs, new_row)
        
      }
      
    } else {
      
      # record accession and the centroid for that accession
      accession <- centroids$V9[1]
      centroid <- centroids$V9[1]
      distance <- as.numeric(all_seqs[all_seqs$Sequence_Name==centroid, "Distance_Score"])
      cluster_size <- as.numeric(all_seqs[all_seqs$Sequence_Name==centroid, "Cluster_Size"])
      month <- i
      cluster <- all_seqs[all_seqs$Sequence_Name==centroid, "Cluster"]
      
      # construct a new row for the all_seqs dataframe
      new_row <- c(accession, distance, cluster_size, month, cluster)
      
      # bind to table
      all_seqs <- rbind(all_seqs, new_row)
      
    }
    
    all_seqs <- all_seqs[order(all_seqs$Distance_Score),]
    rownames(all_seqs) <- NULL
    
    
    # bring these new data into the NCBI metadata
    for (j in 1:nrow(all_seqs)){
      
      accession <- all_seqs$Sequence_Name[j]
      
      if (!("Distance_Score" %in% colnames(metadata))){
        
        # create new columns to fill
        metadata$Distance_Score <- 0.0
        metadata$Cluster_Size <- 0
        metadata$Month_Cluster <- NA
        
      }
      
      # fill in distmat data in the NCBI metadata
      metadata[metadata$Accession==accession, 
               "Distance_Score"] <- as.numeric(all_seqs$Distance_Score[j])
      metadata[metadata$Accession==accession, 
               "Cluster_Size"] <- as.numeric(all_seqs$Cluster_Size[j])
      metadata[metadata$Accession==accession, 
               "Month_Cluster"] <- paste(all_seqs$Month[j], all_seqs$Cluster[j], sep = "_")
      
    }
    
  }
  
  return(metadata)
  
})

# create a function to normalize metadata
normalize_meta <- compiler::cmpfun(function(high_dist_seqs, high_dist_meta){
  
  high_dist_seqs <- high_dist_seqs[names(high_dist_seqs) %in% high_dist_meta$Accession]
  return(high_dist_seqs)
})

# candidate-identifying function to tie it all together
find_candidates <- compiler::cmpfun(function(metadata, stringency){
  
  # creating list of year-month combinations to loop through
  yearmonths <- str_remove_all(list.files(path = ".", pattern = "*-dist-matrix.csv"), "-dist-matrix.csv")
  
  # run the metadata function in a parallelized fashion
  metadata <- process_meta(yearmonths, metadata)
  
  # check to make sure metadata has clustering information in it
  stopifnot(nrow(metadata[!is.na(metadata$Month_Cluster),])>0)
  metadata <- metadata[metadata$Distance_Score>0,] ; rownames(metadata) <- NULL
  
  # make an empty data frame to store high distance FASTA sequences
  high_dist_seqs <- DNAStringSet()
  
  # bring in FASTAs
  fastas <- list.files(path = ".", pattern = "*-cluster-seqs*")
  for (fa in fastas){
    
    # bring in FASTA
    cluster_seqs <- readDNAStringSet(fa)
    
    # add that FASTA to a growing FASTA object of all hits, which will be exported
    # after normalizing later
    high_dist_seqs <- c(high_dist_seqs, cluster_seqs) ; remove(cluster_seqs)
    
  }
  
  # define retention threshold based on the data
  retention_threshold = quantile(metadata$Distance_Score, seq(0, 1, 0.001))[stringency]
  
  # plot distribution of summed weighted distances
  pdf("distance_distribution.pdf", width = 7, height = 5.5)
  hist_vector <-hist(metadata$Distance_Score, freq = FALSE, col = "lightblue",
                     xlab = "Distance Score", ylab = "Frequency", 
                     main = "Frequency Distribution of Nucleotide Distances")
  if (nrow(metadata) > 1){
    lines(density(metadata$Distance_Score), col = "darkblue", lwd = 2)
  }
  abline(v = retention_threshold, col = "red", lwd = 3)
  text(x = retention_threshold+5, y = (max(hist_vector$counts)/2), adj = 0,
       labels = paste("Retention Threshold:\n", retention_threshold ))
  dev.off()
  
  # normalizing year-month by down-prioritizing branches that are not exceptionally
  # long compared to all other year-months. I do this here by retaining only sequences that
  # are in clusters above the 90% quantile of distances, which pulls out true outliers.
  high_dist_meta <- subset(metadata, Distance_Score >= retention_threshold)
  
  # run the function
  high_dist_seqs <- normalize_meta(high_dist_seqs, high_dist_meta)
  
  # check here to prevent silent errors
  stopifnot(as.numeric(length(high_dist_seqs)) == nrow(high_dist_meta))
  
  # writing candidate FASTA of high-distance sequences that passed normalization and 
  # filtering
  writeXStringSet(high_dist_seqs, "high_distance_candidates.fasta")
  
  # writing candidate metadata, which combines vsearch --clust_fast results with 
  # NCBI metadata and iqTree branch lengths
  high_dist_meta <- high_dist_meta[order(high_dist_meta$Distance_Score, decreasing = T),]
  write.table(high_dist_meta, "high_distance_candidates.tsv", row.names = F, quote = F, sep = "\t")
  
})

# run the candidate-finder
find_candidates(metadata, stringency)

File: bin/report-high-dist-candidates.py
#!/usr/bin/env python3

"""
REPORT EVOLUTIONARILY ADVANCED SEQUENCES FOR ADDITIONAL SCRUTINY
----------------------------------------------------------------

This script collates a large number of clustering, genetic distance,
and sequence files to identify sequences that are suitable for
additional review. To do so efficiently, it uses the Polars
library to represent data frames and strictly separates file
reading and writing from CPU-intensive computations. See function
docstrings below for more granular explanations.
----------------------------------------------------------------
"""


# bring in the modules used in this namespace
import argparse
import asyncio
import os
import subprocess
import sys
from dataclasses import dataclass
from glob import glob
from typing import Tuple

import numpy
import polars as pl
import seaborn  # type: ignore
from loguru import logger
from matplotlib import pyplot
from pydantic import validate_call


@validate_call
def parse_command_line_args() -> Tuple[str, str, str, str]:
    """Parse command line arguments."""
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--sequences",
        "-s",
        default="filtered-to-geography.fasta",
        type=str,
        help="The name of the database FASTA file to filter for candidates.",
    )
    parser.add_argument(
        "--metadata",
        "-m",
        default="filtered-to-geography.arrow",
        type=str,
        help="The name of the database metadata file to join other files to.",
    )
    parser.add_argument(
        "--stringency",
        "-str",
        type=str,
        required=True,
        help="The chosen level of strictness for retaining candidates.",
    )
    parser.add_argument(
        "--workingdir",
        "-wd",
        default=".",
        type=str,
        help="Directory where the script should search for files.",
    )
    args = parser.parse_args()
    return args.metadata, args.sequences, args.stringency, args.workingdir


@logger.catch(reraise=True)
@validate_call
async def quantify_stringency(stringency: str) -> int:
    """
    Define strictness level as a quantile.

    Args:
    stringency: a string supplied by the user or the pipeline specifying
    the level of stringency to apply when choosing which candidates to
    retain.

    Returns:
    Integer representing a quantile to apply out of 1,000. This quantile
    will be used to select a retention threshold base on the true
    distribution of distance scores in the data in question.
    """

    if stringency == "strict":
        strict_quant = 995
    elif stringency == "intermediate":
        strict_quant = 990
    elif stringency == "lenient":
        strict_quant = 980
    else:
        strict_quant = 995

    logger.info(
        "Stringency of {} quantified to the {}th quantile.", stringency, strict_quant
    )

    return strict_quant


@logger.opt(lazy=True).catch(reraise=True)
async def visualize_distance_scores(metadata: pl.DataFrame, threshold: float) -> None:
    """
    This function uses Matplotlib and Seaborn to visualize the distribution
    of distance scores in the provided metadata. It also plots the retention
    threshold that was computed based on the data. The graphic is then written
    to a PDF file called "distance_score_distribution.pdf".

    Args:
    metadata - a Polars DataFrame with a column called "Distance Score"
    threshold - a floating point value specifying the distance score retention threshold

    Returns:
    None
    """

    # retrieve the distance scores from the metadata dateframe
    metadata_pd = metadata.to_pandas()
    assert "Distance Score" in metadata_pd.columns
    assert metadata_pd["Distance Score"].nunique() > 1

    # Set the style and size of the plot and generate the histogram
    pyplot.figure(figsize=(7, 5.5))
    seaborn.set_style("whitegrid")
    seaborn.histplot(
        metadata_pd["Distance Score"], kde=True, color="lightblue", element="step"
    )
    max_count = max(pyplot.hist(metadata_pd["Distance Score"], bins=10, alpha=0)[0])

    # Add a vertical line at the retention threshold
    pyplot.axvline(x=threshold, color="red", lw=3)

    # Add a text label for the retention threshold
    pyplot.text(threshold + 20, max_count / 2, f"Retention Threshold:\n{threshold}")

    # Add labels
    pyplot.xlabel("Distance Score")
    pyplot.ylabel("Frequency")
    pyplot.title("Frequency Distribution of Nucleotide Distances")

    # Save the plot to a file
    pyplot.savefig("distance_score_distribution.pdf")


@dataclass
class ClusterTables:
    """
    In-memory tables of information for all clusters/accessions, depending
    on whether preclustering was performed
    """

    full_meta: pl.DataFrame
    dist_scores: pl.DataFrame
    cluster_meta: pl.DataFrame
    whether_preclust: bool


@logger.opt(lazy=True).catch(reraise=True)
@validate_call
async def read_metadata_files(
    metadata_filename: str, yearmonths: list, workingdir: str
) -> ClusterTables:
    """
    Metadata file reading is relegated to this function so that other functions
    can be CPU-bound alone instead of interleaving IO and additional computations.
    The dataframes returned by this function have been organized so that it will
    be trivial to use joins with them downstream.

    Args:
    yearmonths: a list of strings that look like "2021-01"
    workingdir: the working directory. By default, it is the current `pwd` working directory.

    Returns:
    A tuple of three pl.dataframes, one of which is the full database metadata, one
    of which records the distance scores for each accession, and one of which records
    clustering metadata for each accession. These metadata include 1) whether an accession
    is a "hit" or a "centroid," 2) the cluster index for the month each accession was
    collected it, 3) the size of each cluster, and 4) the sequence accessions themselves.
    """

    # start by checking if there are any VSEARCH clustering metadata files
    precluster = len(glob("*.uc")) > 0
    logger.info("Whether preclustering was performed: {}", precluster)

    # prepare to compile distances with an empty dataframe
    dist_scores = pl.DataFrame(
        {"Accession": None, "Distance Score": None},
        schema={"Accession": pl.Utf8, "Distance Score": pl.Float64},
    ).clear()

    # read full input metadata
    metadata = pl.read_ipc(f"{workingdir}/{metadata_filename}", use_pyarrow=True)

    # conditionally prepare to compile VSEARCH cluster metadata
    if precluster:
        cluster_meta = (
            pl.read_csv(
                f"{workingdir}/{yearmonths[0]}-clusters.uc",
                separator="\t",
                has_header=False,
                columns=[0, 1, 2, 8],
                new_columns=["Type", "Index", "Size", "Accession"],
                n_rows=1,
            )
            .with_columns(pl.lit("2023-08").alias("Month"))
            .clear()
        )
    else:
        cluster_meta = metadata.select("Accession")

    # go through each month of data and bring in its various files
    for yearmonth in yearmonths:
        distmat = pl.read_csv(f"{workingdir}/{yearmonth}-dist-matrix.csv")

        # sum up distance score for each accession and prep to vstack on
        # the distance score data frame, which we will use in some joins
        # in a later, non-IO-bound function.
        assert "Sequence_Name" in distmat.columns
        samples = [sample for sample in distmat.columns if sample != "Sequence_Name"]
        sample_size = len(samples)
        sums = [distmat[sample].sum() for sample in samples]
        logger.info(
            "Found {} sequences for {}, with a median distance score of {}",
            sample_size,
            yearmonth,
            (numpy.median(sums) / sample_size),
        )
        distmat = distmat.with_columns(pl.Series(sums).alias("Distance Score")).rename(
            {"Sequence_Name": "Accession"}
        )

        # continue if preclustering was not performed
        if not precluster:
            distmat = (
                distmat.with_columns(pl.lit(sample_size).alias("Sample Size"))
                .with_columns(
                    (pl.col("Distance Score") / pl.col("Sample Size")).alias(
                        "Distance Score"
                    ),
                )
                .select(pl.col(["Accession", "Distance Score"]))
            )

            dist_scores.vstack(distmat, in_place=True)
            continue

        dist_scores.vstack(
            distmat.select(pl.col(["Accession", "Distance Score"])), in_place=True
        )

        # Now, bring in clustering metadata and do the same kind of thing
        cluster_table = (
            pl.read_csv(
                f"{workingdir}/{yearmonth}-clusters.uc",
                separator="\t",
                has_header=False,
                columns=[0, 1, 2, 8],
                new_columns=["Type", "Index", "Size", "Accession"],
            )
            .with_columns(pl.lit(yearmonth).alias("Month"))
            .filter(pl.col("Type") != "S")
        )

        # correct any corrupted rows so the indices can be properly typed
        if cluster_table.dtypes[1] == pl.Utf8:
            cluster_table = cluster_table.filter(
                (pl.col("Type") == "C")
                | (pl.col("Type") == "H") & (pl.col("Index") != "*")
            ).with_columns(pl.col("Index").cast(pl.Int64))

        # vstack (i.e., append) them onto the growing metadata dataframes
        cluster_meta.vstack(cluster_table, in_place=True)

    return ClusterTables(
        full_meta=metadata,
        dist_scores=dist_scores,
        cluster_meta=cluster_meta,
        whether_preclust=precluster,
    )


@logger.opt(lazy=True).catch(reraise=True)
async def collate_metadata(
    cluster_tables: ClusterTables,
    stringency: int,
) -> tuple[pl.DataFrame, pl.DataFrame]:
    """
    This function takes the dataframes read by read_metadata_files and
    runs a number of computations on them. The thrust of all these
    computations is to join them together with the whole-database
    metadata while maximizing helpful information for each candidate
    cluster. Notably, this function handles the CPU-bound tasks that
    are required with the provided metadata.

    Future updates may implement multiprocessing for some of the slower
    steps here. They may also implement LazyFrames instead of dataframes
    to handle very large metadata files.

    Args (parsed from the command line):
    metadata - a pl.dataframe that contains the whole-dataset metadata
    dist_scores - a pl.dataframe that contains distance scores for each
                cluster centroid
    cluster_meta - a pl.dataframe that contains metadata from the
                clustering algorithm.
    """

    # unpack required inputs
    metadata = cluster_tables.full_meta
    dist_scores = cluster_tables.dist_scores
    cluster_meta = cluster_tables.cluster_meta

    # join distance scores onto the cluster metadata, which will add
    # scores to the centroids from each cluster
    assert "Accession" in cluster_meta.columns
    cluster_meta = cluster_meta.join(dist_scores, on="Accession", how="left")

    # Expand cluster centroid distance to all cluster members if preclustering was performed
    if cluster_tables.whether_preclust:
        # merge the month and cluster index columns to make a unique identifier
        cluster_meta = cluster_meta.with_columns(
            [pl.format("{}_{}", "Month", "Index").alias("Month Index")]
        )

        # Use a series of Polars expressions to:
        # 1 - replace sequence length entries in "H" entries with the cluster
        # size entries from the associated "C" entries.
        # 2 - Spread the distance scores from each centroid to all the hit
        # sequences in their clusters, and
        # 3 - select only the cluster size, distance, accession, and month
        # columns for joining with the big metadata
        assert "Type" in cluster_meta.columns
        assert "Month Index" in cluster_meta.columns
        assert "Size" in cluster_meta.columns
        assert "Distance Score" in cluster_meta.columns
        centroid_data = cluster_meta.filter(pl.col("Type") == "C").select(
            ["Month Index", "Size", "Distance Score"]
        )
        cluster_meta = (
            cluster_meta.join(
                centroid_data, on="Month Index", how="left", suffix="_right"
            )
            .select(["Accession", "Month Index", "Size_right", "Distance Score_right"])
            .rename(
                {"Size_right": "Cluster Size", "Distance Score_right": "Distance Score"}
            )
        )

        # filter out any clusters with only one entry (this shouldn't
        # be necessary when run in the context of the pipeline, but
        # we include it here just to be safe.)
        cluster_meta = (
            cluster_meta.group_by("Month Index")
            .agg(pl.col("Month Index").count().alias("count"))
            .join(cluster_meta, on="Month Index", how="left")
            .filter(pl.col("count") > 1)
            .select(cluster_meta.columns)
        )

    # join with the big metadata
    high_dist_meta = metadata.join(cluster_meta, on="Accession", how="left")

    # sort the new high distance metadata by distance score
    # in descending order and remove any null rows
    high_dist_meta = (
        high_dist_meta.sort(by="Distance Score", descending=True, nulls_last=True)
        .filter(pl.col("Distance Score").is_not_null())
        .unique()
    )

    # And finally, filter down to distances above the retention threshold while
    # also visualizing the distance score distribution
    retention_threshold = float(
        numpy.quantile(high_dist_meta["Distance Score"], (stringency / 1000))
    )
    try:
        await visualize_distance_scores(high_dist_meta, retention_threshold)
    except Exception as traceback:  # pylint: disable=W0718
        logger.debug("Plotting the distance score distribution failed: {}", traceback)
    logger.info("Distance score retention threshold: {}", retention_threshold)
    logger.info(
        "Max-observed distance score: {}",
        max(high_dist_meta.select("Distance Score").to_series().to_list()),
    )
    high_dist_meta = high_dist_meta.filter(
        pl.col("Distance Score") >= retention_threshold
    )

    # separate out the accessions
    accessions = high_dist_meta.select("Accession")

    return high_dist_meta, accessions


async def main():
    """
    Main function. Main ties together all the above functions if
    they are being invoked as a script. It also works to "glue"
    in a subprocess command using Seqkit to handle FASTA filtering
    (Seqkit, written in Go, is much faster at filtering large
    numbers of FASTA records than Python).
    """

    # initialize logger
    logger.remove()
    logger.add(sys.stderr, backtrace=True, diagnose=True, colorize=True, enqueue=True)

    # retrieve file paths and settings from keyword command line arguments
    metadata_name, fasta_path, stringency, workingdir = parse_command_line_args()

    # check that the provided input files exist
    assert os.path.isfile(
        metadata_name
    ), "Provided metadata path does not point to a file that exists."
    assert os.path.isfile(
        fasta_path
    ), "Provided FASTA path does not point to a file that exists."
    assert os.path.isdir(workingdir), "Provided working directory path does not exist."

    # Retrieve a quantile to use to define a retention threshold downstream
    strict_quant = await quantify_stringency(stringency)
    logger.info("Selected quantile stringency: {}", strict_quant)

    # List all files in current directory that end with '-dist-matrix.csv'
    logger.info("Collecting available distance matrix files.")
    files = [
        file for file in os.listdir(workingdir) if file.endswith("-dist-matrix.csv")
    ]

    # Remove '-dist-matrix.csv' from each file name
    yearmonths = [file.replace("-dist-matrix.csv", "") for file in files]

    # retrieve all candidate metadata
    logger.info("Reading candidate metadata files.")
    cluster_tables = await read_metadata_files(metadata_name, yearmonths, workingdir)

    # pile up all metadata from clustering, distance-calling, and the original database
    logger.info(
        "Piling up all metadata from clustering, distance-calling, and the original dataset"
    )
    high_dist_meta, accessions = await collate_metadata(
        cluster_tables,
        strict_quant,
    )

    # Use an assertion to spare the computer some effort if no candidates were found
    assert accessions.shape[0] > 0, "No candidate sequences identified."

    # write the metadata and the accessions
    logger.info("Writing high-distance metadata and accession.")
    high_dist_meta.write_csv(
        f"{workingdir}/high_distance_candidates.tsv", separator="\t"
    )
    accessions.write_csv(
        f"{workingdir}/high-dist-accessions.txt", include_header=False, separator="\t"
    )

    # Use Seqkit to separate out high distance sequences based on the metadata
    logger.info("Filtering FASTA records to high distance accessions only.")
    cmd = [
        "seqkit",
        "grep",
        "-f",
        "high-dist-accessions.txt",
        fasta_path,
        "-o",
        f"{workingdir}/high_distance_candidates.fasta",
    ]
    subprocess.run(cmd, check=True)

    # remove the accessions file that is now unnecessary
    os.remove("high-dist-accessions.txt")


if __name__ == "__main__":
    asyncio.run(main())
